config/config.go

```diff
@@ -0,0 +1,459 @@
+                    'h//yy///sd+///ssoooooooso:..'   '+-dys'-h+          
+                    'y//ys///+s///+hoooooooso'h++o+'.-hydhyyshy'         
+                     y//++////////ssooooooos 'y+/sso+++syyyyyhyo  ''     
+                     o+///////+s//yoooooooh:+oo+:/:::::::+++ooshss++o    
+                     's///////yh//ysoooooodo/:::::::::::::::::::/sss+    
+                   //:/s//////hyo/+hoooosh+:::::::::::::::::::::::/y.    
+                   s/+os////++hoh//ysoosh/::::::::::::+o///+o/:::::/h'   
+                   y////////hyhosy//yssh::::::::::::/o-     ':s/::::so.  
+                   +o//////+yshyosy+/sd/::::::::::::y      '+hyy::::o+:o.
+                   'y//////+hoodyooy+/+ssoo+/+oooo::s      .mNsy::::s::dy
+///.                /+//////ssoosyooss+////+++//+y::s:        ++:/ssh .ss
+++/oo/.    o+/.      s////+yosyooooooossso++++//y/:::+o/-''-/o+s++syy-// 
+ +o///o+/. oo/+o/- -soh+///yysyhoooooooooooosyoso:::::::/++/::/s....o:   
+  .+o////+o/os+//+o/y++so///+ssssooooooooossh/-::/:::::::::::::o/oo//    
+    .+o+////+shy+///oy+/oso////++ossssssso+os+++/::::::::::::::+oy+/:    
+      '-+o+////+o+///////ossssoo+++++oooooo/:::::::::o/::::::::/h/::/    
+       s++sys+//yysssso+oysssm///+++////::::::::::::/s:::/+:::/y'''      
+       :+o+++oo/+ssssoosyyhyod:::::::::::::::::::::+o:::+h++:/s'         
+        '-+++/////+++shyyooooys::::::::::::::::::::s++++sooh+s'          
+           './oysoo///osssssssho::::::::::::::::::::/:::o+/so'           
+              ss+++//////////+oys/:::::::::::::::::::::::+o:             
+              '.:/+++++++++//-.'.soo/o::::::::::::::::/oo:'              
+                   ''''''        h/-./ss+++///////++oo/-'                
+                                 oo++/.''o+/y////::-.                    
+                                         ://-                            
+        ____  _____   _                                          
+       |_   \|_   _| (_)                                         
+         |   \ | |   __   _ .--.  _   __  ,--.   _ .--.   ,--.   
+         | |\ \| |  [  | [ '/''\][ \ [  ]''_\ : [ '.-. | ''_\ :  
+        _| |_\   |_  | |  | |     \ \/ / // | |, | | | | // | |, 
+       |_____|\____|[___][___]     \__/  \'-;__/[___||__]\'-;__/ 
+
+`
+
+// Option must be a pointer to struct.
+type Option interface{}
+
+// Plugin is for plugins to collect configurations
+type Plugin interface {
+	// Name returns plugin name.
+	Name() string
+	// Configure configures nirvana config via current options.
+	Configure(cfg *nirvana.Config) error
+}
+
+// NirvanaStarter is a nirvana starter.
+type NirvanaStarter interface {
+	// EnablePlugin enables plugins.
+	EnablePlugin(plugins ...Plugin) NirvanaStarter
+	// AddOption will fill up options from config/ENV/flags after executing.
+	AddOption(prefix string, options ...Option) NirvanaStarter
+	// Add adds a field by key
+	Add(pointer interface{}, key string, shortFlag string, desc string) NirvanaStarter
```

Comment:
`Add` confuse me, add what? It SHOULDE be AddFlag
and the first arg pointer is not required.
should we provide `AddFlag` and `AddFlagP` just like cobra?

Reply:
How can I add a persistent flag?

Reply:
This is used to add to internal store?

Reply:
`Add` needs more comments and it SHOULD NOT be `AddFlag`.

`Add` add a field to receive configuration from `Flag/ENV/ConfigFile`.  That means the key is like `nirvana.ip` instead of `nirvana-ip`. Same approach to get configuration is `config.GetString("nirvana.ip")`.

Also, you should not add a persistent flag.